File: js/006 ZigZag Conversion/ZigZagConversion.js

/**
 * 本题题意：给出一个字符串，该字符串会通过Z字形排列，然后给出Z字形排列的顺序读取结果。
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
  if (numRows == 1) {
    return s;
  }
  let x = 2 * numRows - 2;
  let sbs = [];
  for (let i = 0; i < numRows; i++) {
    sbs[i] = '';
  }

  let isFront = true;
  for (let i = 0, j = 0; i < s.length; i++) {
    if ((i % x) == 0) {
      isFront = true;
    } else if ((i % x) == numRows - 1) {
      isFront = false;
    }
    print(j, i, x, (i % x))
    sbs[isFront ? j++ : j--] += s[i];
  }
  return sbs.join('');
};

File: py/006 ZigZag Conversion/ZigZagConversion.py

# 本题题意：给出一个字符串，该字符串会通过Z字形排列，然后给出Z字形排列的顺序读取结果。

class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows is 1:
            return s
        x = 2 * numRows - 2
        sbs = [[] for i in range(numRows)]
        isFront = True
        j = 0
        for i in range(0, len(s)):
            if int(i % x) == 0:
                isFront = True
            elif int(i % x) == numRows - 1:
                isFront = False;

            sbs[j].append(s[i])
            j += 1 if isFront else -1

        return "".join(["".join(e) for e in sbs])
